waspc/data/Generator/templates/server/src/auth/providers/email/signup.ts

```diff
@@ -2,70 +2,108 @@ import { Request, Response } from 'express';
+    rethrowPossibleAuthError,
 } from "../../utils.js";
 import {
-    createEmailVerificationLink,
+    createEmailVerificationLinkWithToken,
     sendEmailVerificationEmail,
     isEmailResendAllowed,
 } from "./utils.js";
 import { ensureValidEmail, ensureValidPassword, ensurePasswordIsPresent } from "../../validation.js";
 import { GetVerificationEmailContentFn } from './types.js';
 import { validateAndGetAdditionalFields } from '../../utils.js'
+import HttpError from '../../../core/HttpError.js';
 
 export function getSignupRoute({
     fromField,
     clientRoute,
     getVerificationEmailContent,
+    allowUnverifiedLogin,
 }: {
     fromField: EmailFromField;
     clientRoute: string;
     getVerificationEmailContent: GetVerificationEmailContentFn;
+    allowUnverifiedLogin: boolean;
 }) {
     return async function signup(
         req: Request<{ email: string; password: string; }>,
         res: Response,
-    ): Promise<Response<{ success: true } | { success: false; message: string }>> {
-        const userFields = req.body;
-        ensureValidArgs(userFields);
+    ): Promise<Response<{ success: true }> | void> {
+        const fields = req.body;
+        ensureValidArgs(fields);
         
-        userFields.email = userFields.email.toLowerCase();
+        const providerId = createProviderId("email", fields.email);
+        const existingAuthIdentity = await findAuthIdentity(providerId);
 
-        const existingUser  = await findUserBy({ email: userFields.email });
-        // User already exists and is verified - don't leak information
-        if (existingUser && existingUser.isEmailVerified) {
-            await doFakeWork();
-            return res.json({ success: true });
-        } else if (existingUser && !existingUser.isEmailVerified) {
-            if (!isEmailResendAllowed(existingUser, 'emailVerificationSentAt')) {
-                return res.status(400).json({ success: false, message: "Please wait a minute before trying again." });
+        if (existingAuthIdentity && !allowUnverifiedLogin) {
+            const providerData = deserializeAndSanitizeProviderData<'email'>(existingAuthIdentity.providerData);
+
+            // 1a. User already exists and is verified - don't leak information
+            if (providerData.isEmailVerified) {
+                await doFakeWork();
+                return res.json({ success: true });
+            }
+
+            if (!isEmailResendAllowed(providerData, 'emailVerificationSentAt')) {
+                throw new HttpError(400, "Please wait a minute before trying again.");
             }
-            await deleteUser(existingUser);
+
+            try {
+                // 1b. User exists but is not verified - delete the user and create a new one
+                await deleteUserByAuthId(existingAuthIdentity.authId);
+            } catch (e: unknown) {
+                rethrowPossibleAuthError(e);
+            }
+
+        } else if (existingAuthIdentity && allowUnverifiedLogin) {
```

Comment:
Yeah, let's reafctor it.

I get the point with nesting, but what you loose is obvious branching -> instead of reader being very clear about all cases being handled, I have to think quite a bit here to figure out these two really cover all the cases. And it can also be easier in the future to introduce a mistake where these conditions don't perfectly match each other and suddenly they don't cover all the cases any more.